functions/snippets/src/main/java/com/example/functions/SlackSlashCommand.java

```diff
@@ -0,0 +1,172 @@
+import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+public class SlackSlashCommand implements HttpFunction {
+
+  private Kgsearch kgClient;
+  private static String API_KEY;
+  private static String SLACK_SECRET;
+  private static final Logger LOGGER = Logger.getLogger(HelloHttp.class.getName());
+  private SlackSignature.Verifier verifier;
+  private Gson gson = new Gson();
+
+  public SlackSlashCommand() throws IOException, GeneralSecurityException {
+    kgClient = new Kgsearch.Builder(
+        GoogleNetHttpTransport.newTrustedTransport(), new JacksonFactory(), null).build();
+
+    // Read + parse config file
+    Path configPath = Path.of(System.getProperty("user.dir"), "config.json");
+    JsonObject configJson = (new Gson()).fromJson(Files.readString(configPath), JsonObject.class);
+
+    SLACK_SECRET = configJson.get("SLACK_SECRET").getAsString();
+    API_KEY = configJson.get("KG_API_KEY").getAsString();
+
+    verifier = new SlackSignature.Verifier(new SlackSignature.Generator(SLACK_SECRET));
+  }
+
+  private boolean isValidSlackWebhook(HttpRequest request, String requestBody) throws IOException {
+
+    // Check for headers
+    HashMap<String, List<String>> headers = new HashMap(request.getHeaders());
+    if (!headers.containsKey("X-Slack-Request-Timestamp")
+        || !headers.containsKey("X-Slack-Signature")) {
+      return false;
+    }
+
+    return verifier.isValid(
+        headers.get("X-Slack-Request-Timestamp").get(0),
+        requestBody,
+        headers.get("X-Slack-Signature").get(0),
+        1L);
+  }
+
+  private void addPropertyIfPresent(
+      JsonObject target, String targetName, ArrayMap source, String sourceName) {
+    if (source.containsKey(sourceName)) {
+      target.addProperty(targetName, source.get(sourceName).toString());
+    }
+  }
+
+  private String formatSlackMessage(List<Object> kgResults, String query) {
+    JsonObject attachmentJson = new JsonObject();
+    JsonArray attachments = new JsonArray();
+
+    JsonObject responseJson = new JsonObject();
+    responseJson.addProperty("response_type", "in_channel");
+    responseJson.addProperty("text", String.format("Query: %s", query));
+
+    // Extract the first entity from the result list, if any
+    if (kgResults.size() == 0) {
+      attachmentJson.addProperty("text","No results match your query...");
+
+      attachments.add(attachmentJson);
+      responseJson.add("attachments", attachmentJson);
+
+      return gson.toJson(responseJson);
+    }
+
+    ArrayMap entity = (ArrayMap) ((ArrayMap) kgResults.get(0)).get("result");
+
+    // Construct Knowledge Graph response attachment
+    String title = entity.get("name").toString();
+    if (entity.containsKey("description")) {
+      title = String.format("%s: %s", title, entity.get("description").toString());
+    }
```

Comment:
Lots of .toString()'s are often a flag of bad design.  .toString() on an object isn't guaranteed to provide something presentable.